### PR TITLE
Yo 21

### PR DESCRIPTION
**Pull request by ladinu**
# Description

Added carpool creation and carpool retrieval by id functionality. Checkout swagger docs for more info.

### CODE REVIEW

**Comment by ianaldrighetti** [comment]
@ladinu Is this branch different than the yo-carpool-service branch?

**Comment by ladinu** [comment]
No. `yo-carpool-service` and `yo-21` is pretty much the same branch. I merged them back and forth while I was working on the story.

**Comment by ianaldrighetti** [comment]
Oh okay. I just pushed some stuff up into yo-carpool-service not that long ago. 

Also, I forgot to do that outputJSON thing you wanted me to do in the campus controller. So I need to do that.

**Comment by ladinu** [comment]
I forgot I needed `outputJSON` from you. Let me know when that is done because I need to use it in this branch.

File: src/controllers/carpool.ctrl.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        carpoolName, campusName, description, owner)
+        .then((carpool) => {
+          res.send(201, toOutputJSON(carpool));
+        })
+        .catch(errors.CarpoolOwnerNotFoundException,
+          errors.CampusNotFoundException, (err) => {
+          next(new restify.NotAcceptableError(err.message));
+        })
+        .catch(errors.CarpoolExistException, (err) => {
+          next(new restify.ConflictError(err.message));
```
Comment (joe-yeager) [inline src/controllers/carpool.ctrl.ts:45]: Do we have middleware to handle the errors passed to next()? Or where are these errors getting translated to the corresponding status codes?

Edit: Nvm, I just realized these are restify errors......

Comment (ladinu) [inline src/controllers/carpool.ctrl.ts:45]: If you pass an `Error` object to `next()`, restify will automatically format that into a HTTP 500 by default. The `Error` object will get converted into a JSON object. If you use the built-in restify errors, restify will set the appropriate HTTP status code. 

So if I call `next(new restify.ConflictError("foo"))`, the HTTP status would be 409.

Comment (joe-yeager) [inline src/controllers/carpool.ctrl.ts:45]: Yeah, I just realized they were restify errors right before you hit the comment button haha.  For some reason I thought these were custom errors that you defined. I either need more coffee or sleep.

File: src/server.ts [hunk 2]
```diff
@@ -66,6 +66,9 @@ server.get('/flavors', function(req, res, next) {
 });
 
 server.post('/users', Controller.createUser);
+server.get('/carpools', Controller.getCarpools);
+server.post('/campus/create', Controller.createCampus);
+server.get('/campus/list', Controller.getCampusList);
```
Comment (joe-yeager) [inline src/server.ts]: These should probably go in the routes.ts file to be consistent.

**Comment by joe-yeager** [comment]
#peer

**Comment by ianaldrighetti** [comment]
@ladinu I can't do it until I get home, sorry. I tried to do it at work and commit it but the port is blocked when I attempt to push it up.

In the meantime, are you able to merge from yo-carpool-service into this branch here? I added some changes last night in there and I was not aware I should have probably been doing it in yo-21 (I fixed those endpoint routes like you asked and put them in the routes.ts file like Joseph suggested above).

File: src/server.ts (hunk 2 above)
Comment (ladinu) [inline src/server.ts]: Yeah we should remove the routes from here.

**Comment by ladinu** [comment]
@ianaldrighetti I merged `yo-carpool-serice` into this branch few minutes ago. I still don't see the changes you made. Try changing the git port to 80 or 443 instead of 22. I usually clone git repos with http instead of SSH.

**Comment by ianaldrighetti** [comment]
Hmm, okay, strange. I'll check my personal computer around lunch time to see if I forgot to push it up.

Sorry about that.

**Comment by ladinu** [comment]
@ianaldrighetti I just pushed up what you did in `yo-carpool-service`. Everything looks good! 

Only concern is I was not able to create a `Campus` using swagger interface (was able to create campus using PostMan just fine). Can you look into this? 

Also who is working on the integration test for the `CampusController`? I know @phanmn1 is working on the integration tests for the `CampusService`.

**Comment by ianaldrighetti** [comment]
@ladinu Okay great, and yes I can look into that later today.

**Comment by ianaldrighetti** [comment]
#wewerethere

**Comment by phanmn1** [comment]
#washere

**Pull request merged**